torch/csrc/distributed/c10d/init.cpp

```diff
@@ -115,9 +117,62 @@ PyObject* c10d_init(PyObject* _unused) {
               py::arg("op") = ::c10d::ReduceOp::SUM,
               py::call_guard<py::gil_scoped_release>());
 
-  shared_ptr_class_<::c10d::ProcessGroupGloo>(
-      module, "ProcessGroupGloo", processGroup)
-      .def(py::init<const std::shared_ptr<::c10d::Store>&, int, int>());
+  auto processGroupGloo = shared_ptr_class_<::c10d::ProcessGroupGloo>(
+      module, "ProcessGroupGloo", processGroup);
+
+  shared_ptr_class_<::gloo::transport::Device>(processGroupGloo, "Device");
+
+  shared_ptr_class_<::c10d::ProcessGroupGloo::Options>(
+      processGroupGloo, "Options")
+      .def(py::init<>())
+      .def_readwrite("timeout", &::c10d::ProcessGroupGloo::Options::timeout)
+      .def_readwrite("threads", &::c10d::ProcessGroupGloo::Options::threads)
+      .def_readwrite("devices", &::c10d::ProcessGroupGloo::Options::devices);
+
+  processGroupGloo.def_static(
+      "create_tcp_device",
+      [](const std::string& hostname, const std::string& interface)
+          -> std::shared_ptr<::gloo::transport::Device> {
+        ::gloo::transport::tcp::attr attr;
+        if (!hostname.empty()) {
+          attr.hostname = hostname;
+        } else if (!interface.empty()) {
+          attr.iface = interface;
+        } else {
+          // Neither argument is specified; Gloo itself will use the hostname
+          // Nothing specified, default to something useful
+        }
+        return ::gloo::transport::tcp::CreateDevice(attr);
+      },
+      py::arg("hostname") = "",
+      py::arg("interface") = "");
+
+  processGroupGloo
+      .def(py::init<
+           const std::shared_ptr<::c10d::Store>&,
+           int,
+           int,
+           ::c10d::ProcessGroupGloo::Options>())
+      .def(py::init(
+          [](const std::shared_ptr<::c10d::Store>& store, int rank, int size) {
+            ::c10d::ProcessGroupGloo::Options options;
+
+            // By default, use the hostname to resolve the network address to
+            // use. Note: if the hostname does not resolve to an address (e.g.
+            // because of misconfigured /etc/hosts file), this will not work.
+            std::array<char, HOST_NAME_MAX> hostname;
+            auto rv = gethostname(hostname.data(), hostname.size());
+            if (rv != 0) {
+              throw std::system_error(errno, std::system_category());
+            }
```

Comment:
Uh I remember this is what gloo did, but it was a constant source oh headaches of me, because it failed miserably on many of my home machines, which apparently configure their hostname to point to `127.0.0.1` by default. Can we use something like `getInterfaceAddresses()` in THD?

Reply:
THD uses `config.publicAddress`, which it looks like is returned from the TCP init server (so it will always be the external IP, since it could route to the specified TCP address). I'll see if I can copy `getInterfaceAddresses`, as we can't depend on somebody else giving us our address.

Reply:
IIRC in THD all our init methods broadcast all possible interface addresses that a given machine has. This is what `getInterfaceAddresses` returns. Then, they will simply iterate over those, and try to connect to them in order, until one succeeds. It's much more foolproof.

Reply:
That is interesting. I see the multicast and the direct TCP init methods use a slightly different approach but both end up with the routeable address. We could do the same by leveraging the store, but this would be a preamble to ending up with a Gloo device and doing some accept(2)/connect(2) business.

I agree we need something like this to make it work out of the box in the majority of cases, but this is lower priority than some of the other stuff that needs to be done. Would you be OK merging this without the auto detection like done in THD, but with detection and throwing on seeing a loopback address?

Reply:
I added an issue to track. I'd like to merge this to unblock other stuff and then deal with this separately.

Reply:
Sure I’m ok with that. Just leaving it here, because i remember it was pretty annoying for me back then